// db/index.js
var mongodb = require('mongodb')
var server = new mongodb.Server("localhost", 27017, {safe: true})
var workflow = new mongodb.Db('workflow', server, {journal: true})

exports.mongodb = mongodb

exports.dbServer = workflow

exports.Collection = mongodb.Collection

//当数据库连接成功后，需要执行的一些回调
//例如：需要在第一时间缓存好某些数据
exports._cache = []

exports.open = function (callback) {
    if (typeof callback === 'function' && !exports.Client) {
        exports._cache.push(callback)
    }

    if (exports.Client) {
        callback()
        return
    }
    workflow.open(function (error, client) {
        if (!error) {
            exports.Client = client
            //开始建立用户数据库链接
            var userServer = new mongodb.Server("localhost", 27017, {safe: true})
            var user = new mongodb.Db('user', userServer, {journal: true})
            user.open(function (error, client) {
                if (!error) {
                    exports.userClient = client
                    exports.userServer = userServer
                    //将队列中存在的函数列表全部执行一次。
                    exports._cache.forEach(function (fn) {
                        fn()
                    })
                    delete exports._cache
                }
                console.log(!error ? '用户数据库连接成功' : '无法链接用户数据库')
            })
        }
        console.log(!error ? '数据库连接成功' : '无法连接数据库')
    })
}

workflow.on('close', function () {
    console.log('Database connection is disconnected!\t' + new Date().toLocaleTimeString())
})
